libportal/remote.c

```diff
@@ -905,9 +906,74 @@ is_active_remote_desktop_session (XdpSession    *session,
   return XDP_IS_SESSION (session) &&
          session->type == XDP_SESSION_REMOTE_DESKTOP &&
          session->state == XDP_SESSION_ACTIVE &&
+         !session->uses_eis &&
          (session->devices & required_device) != 0;
 }
 
+/**
+ * xdp_session_connect_to_eis
+ * @session: a [class@Session]
+ * @error: return location for a #GError pointer
+ *
+ * Connect this XdpRemoteDesktopSession to an EIS implementation and return the fd.
+ * This fd can be passed into ei_setup_backend_fd(). See the libei
+ * documentation for details.
+ *
+ * This call must be issued before xdp_session_start(). If successful, all input
+ * event emulation must be handled via the EIS connection and calls to
+ * xdp_session_pointer_motion() etc. are silently ignored.
+ *
+ * Returns: the file descriptor to the EIS implementation
+ */
+int
+xdp_session_connect_to_eis (XdpSession  *session,
+                            GError     **error)
+{
+  XdpPortal *portal = session->portal;
+  GVariantBuilder options;
+  g_autoptr(GVariant) ret = NULL;
+  g_autoptr(GUnixFDList) fd_list = NULL;
+  int fd_out = -1;
+
+  if (portal->remote_desktop_interface_version < 2)
+      g_set_error (error, G_IO_ERROR, G_IO_ERROR_NOT_SUPPORTED, "Not supported by the portal interface");
+  else if (session->type != XDP_SESSION_REMOTE_DESKTOP)
+      g_set_error (error, G_IO_ERROR, G_IO_ERROR_INVALID_ARGUMENT, "Session is not a Remote Desktop session");
+  else if (xdp_session_get_session_state (session) != XDP_SESSION_ACTIVE)
+      g_set_error (error, G_IO_ERROR, G_IO_ERROR_INVALID_ARGUMENT, "Session has not been started");
+  else if (session->uses_eis)
+      g_set_error (error, G_IO_ERROR, G_IO_ERROR_INVALID_ARGUMENT, "Session is already connected to EIS");
+
+  if (error != NULL && *error != NULL)
+      return -1;
```

Comment:
Sorry to be such a pain. I think these checks misbehave when someone calls `xdp_session_connect_to_eis (session, NULL);`. If NULL is passed, then this code will proceed to call `g_dbus_connection_call_with_unix_fd_list_sync()` regardless of having an error.

I think you have to either make all checks be something like this:

```c
if (portal->foo < bar)
  {
    g_set_error (error, <some error>);
    return -1;
  }
```

Or use an auxiliary variable:

```c
g_autoptr(GError) local_error = NULL;

if (portal->foo < bar)
  g_set_error (&local_error, <some error>);
else if (portal->frob < nicate)
  g_set_error (&local_error, <some error>);

if (local_error != NULL)
  {
    g_propagate_error (error, g_steal_pointer (&local_error);
    return -1;
  }
```

I don't have any strong preference, I'm slightly tending towards the first example as it would avoid allocating and deallocating a GError, but either should be fine.

(Also style nit: all `g_set_error`s and the `return -1;` are one indentation level off)

Reply:
oh, right, good catch. I tried to be too smart for my own good (had the `{...return -1; }` first and then tried to polish it - which also explains the indentation.

Done in the first style now, thanks for catching this. Unfortunately the python tests didn't because the PyGI bindings always pass a GError.